tests/testsuite/git_state.rs

```diff
@@ -0,0 +1,177 @@
+use std::ffi::OsStr;
+use std::fs::OpenOptions;
+use std::io::{self, Error, ErrorKind, Write};
+use std::process::{Command, Stdio};
+
+#[test]
+#[ignore]
+fn shows_rebasing() -> io::Result<()> {
+    let repo_dir = create_repo_with_conflict()?;
+    let path = path_str(&repo_dir)?;
+
+    run_git_cmd(&["rebase", "other-branch"], Some(path), false)?;
+
+    let output = common::render_module("git_state")
+        .current_dir(path)
+        .output()?;
+    let text = String::from_utf8(output.stdout).unwrap();
+    assert!(text.contains("REBASING 1/1"));
+
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn shows_merging() -> io::Result<()> {
+    let repo_dir = create_repo_with_conflict()?;
+    let path = path_str(&repo_dir)?;
+
+    run_git_cmd(&["merge", "other-branch"], Some(path), false)?;
+
+    let output = common::render_module("git_state")
+        .current_dir(path)
+        .output()?;
+    let text = String::from_utf8(output.stdout).unwrap();
+    assert!(text.contains("MERGING"));
+
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn shows_cherry_picking() -> io::Result<()> {
+    let repo_dir = create_repo_with_conflict()?;
+    let path = path_str(&repo_dir)?;
+
+    run_git_cmd(&["cherry-pick", "other-branch"], Some(path), false)?;
+
+    let output = common::render_module("git_state")
+        .current_dir(path)
+        .output()?;
+    let text = String::from_utf8(output.stdout).unwrap();
+    assert!(text.contains("CHERRY-PICKING"));
+
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn shows_bisecting() -> io::Result<()> {
+    let repo_dir = create_repo_with_conflict()?;
+    let path = path_str(&repo_dir)?;
+
+    run_git_cmd(&["bisect", "start"], Some(path), false)?;
+
+    let output = common::render_module("git_state")
+        .current_dir(path)
+        .output()?;
+    let text = String::from_utf8(output.stdout).unwrap();
+    assert!(text.contains("BISECTING"));
+
+    Ok(())
+}
+
+#[test]
+#[ignore]
+fn shows_reverting() -> io::Result<()> {
+    let repo_dir = create_repo_with_conflict()?;
+    let path = path_str(&repo_dir)?;
+
+    run_git_cmd(&["revert", "--no-commit", "HEAD~1"], Some(path), false)?;
+
+    let output = common::render_module("git_state")
+        .current_dir(path)
+        .output()?;
+    let text = String::from_utf8(output.stdout).unwrap();
+    assert!(text.contains("REVERTING"));
+
+    Ok(())
+}
+
+fn run_git_cmd<A, S>(args: A, dir: Option<&str>, expect_ok: bool) -> io::Result<()>
+where
+    A: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new("git");
+    command
+        .args(args)
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .stdin(Stdio::null());
+
+    if let Some(dir) = dir {
+        command.current_dir(dir);
+    }
+
+    let status = command.status()?;
+
+    if expect_ok && !status.success() {
+        Err(Error::from(ErrorKind::Other))
+    } else {
+        Ok(())
+    }
+}
+
+fn create_repo_with_conflict() -> io::Result<tempfile::TempDir> {
```

Comment:
Well. That is one monster of a function, and you've written it beautifully. 10/10, would read again.